### PR TITLE
Find ElastiCube dependencies

### PR DESCRIPTION
**Pull request by ricardolsmendes**
**- What I did**
Added the `addons.ElastiCubeDependencyFinder` class, the core of a new feature for the Sisense Connector CLI that is going to [allow users to easily find and display the specific fields and filters that depend on a given ElastiCube data source, table, and/or column](https://github.com/GoogleCloudPlatform/datacatalog-connectors-bi/projects/1#card-68317635).

**- How I did it**
Added the `addons.ElastiCubeDependencyFinder` class and its unit tests as well.

**- How to verify it**
Run the unit tests.

**- Description for the changelog**
Added the `addons.ElastiCubeDependencyFinder` class.

PS 1: This PR is part of the effort to deliver #70.
PS 2: The CLI changes and related docs are already done but will be tackled in a separate PR to make the code review easier.

### CODE REVIEW

File: google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+    @classmethod
+    def __filter_table_column_matching_tags(cls, table: str, column: str,
+                                            tags: List[Tag]) -> List[Tag]:
+        """Filter JAQL-related Tags which `table` and/or `column` fields match
+        the provided values.
+
+        If both `table` and `column` args are provided, a given Tag's
+        corresponding fields must match both values in order to added to the
+        resulting list.
```
Comment (mesmacosta) [inline google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py:279]: small typo, in order to **be** added to the resulting list.

File: google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+        Returns:
+            list: The filtered Tags.
+        """
+        jaql_tags = cls.filter_jaql_tags(tags)
+        table_matching_tags = cls.__filter_table_matching_tags(
+            table, jaql_tags)
+        column_matching_tags = cls.__filter_column_matching_tags(
```
Comment (mesmacosta) [inline google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py:287]: The method logic, docstring and the method signature don't match:

1 -  Signature  
`__filter_table_column_matching_tags(cls, table: str, column: str...`  
table and column are required method args.

2 - Method logic  
```
        column_matching_tags = cls.__filter_column_matching_tags(
            column, table_matching_tags if table and column else tags)
```
Here table and column can be optional, so I'd suggest to either:
- Change method signature and use Optional[str], indicating that method args can be optional
- Change method logic to assume the args are always required

Comment (mesmacosta) [reply google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py:287]: There are other scenarios, I guess Optional is not really being used, so this could be addressed later in a separate PR. Since typing is being used in the codebase, would be nice to also explicitly type what's Optional and what is not...

This one caught my attention, because changes the method logic, it's not just a simple check like the other methods.

**Review by mesmacosta — APPROVED** [review]
just some notes, but LGTM!

File: google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py (hunk 1 above)
Comment (ricardolsmendes) [reply google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py]: Fixed, thanks for catching this.

File: google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py (hunk 2 above)
Comment (ricardolsmendes) [reply google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py:293]: The `table` and `column` args are actually optional, so I did small changes In order to declare them as named args with `None` as their default value wherever it makes sense. I decided on this approach because `Optional` is a type hint and ignoring it does not incur programming errors AFAIK (please correct me if I'm wrong).

**Comment by ricardolsmendes** [comment]
Hey @mesmacosta, although you have already approved the PR, I'll kindly ask you to have a look at the `table` and `column` named args before merging it. Please let me know if the current approach makes sense to you. Thanks!

File: google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py (hunk 2 above)
Comment (mesmacosta) [reply google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py:293]: That is fine by me, using the default value makes it clear that the parameters are optional. Before my point was that it was confusing looking at the method args I'd think the args were required.

About type hints not incurring programming errors, you are right. But you can use the default args together with the type hints.

And you are already using those: `from typing import Dict, List, Optional, Tuple, Union`

If you are applying type hints, I'd prefer to use them all the way, we can add `mypy` to the build, and it would catch those typing errors.

Even `Optional` typing will be caught by it, let's say you mark an arg as Optional, and you don't handle the None value in the method, `mypy` will break the build and your IDE will also warn you. I've been using Optional heavily and it's helpful.

So to sum up, I'm ok if you prefer not to use it for now, but I'd suggest to use the default args as you are doing and also add Optional to them.

**Review by mesmacosta — APPROVED** [review]
just added a note to it, LGTM!

File: google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py (hunk 2 above)
Comment (ricardolsmendes) [reply google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py:298]: Thanks for clarifying, @mesmacosta. I've added `Optional[str]` to the function args where it applies in the `ElastiCubeDependencyFinder` class. I intend to let the new code be as ready as possible for `mypy` 😃.

In regards to adding `mypy` checks to the CI, I'd rather tackle this in a separate MR because it could result in more changes to the whole code.

Comment (mesmacosta) [reply google-datacatalog-sisense-connector/src/google/datacatalog_connectors/sisense/addons/elasticube_dependency_finder.py:298]: yup! makes sense, will merge the MR then, and `mypy` can be handled in a separate MR.

**Review by mesmacosta — APPROVED** [review]
